### PR TITLE
Implement smartErase functionality [NXTFLASHER-581]

### PR DESCRIPTION
**Pull request by jneis-at-hilscher**
Implemented smartErase functionality which will try to optimise the erase times by using larger erase operations where multiple sectors in a row need to be erased.

### CODE REVIEW

File: templates/flasher.lua [hunk 1]
```diff
@@ -960,6 +961,80 @@ end
...
+		end
+		
+		ulEndOffset = ulEraseEnd
+	else
+		ulEndOffset = ulDeviceOffset + ulSize
+	end
+	
+	print(string.format("Area:  [0x%08x, 0x%08x[", ulDeviceOffset, ulEndOffset))
+	print("Checking if the area is already empty")
+	fIsErased = isErased(tPlugin, aAttr, ulDeviceOffset, ulEndOffset, fnCallbackMessage, fnCallbackProgress)
```
Comment (dizzydevil) [inline templates/flasher.lua:1013]: We might remove this check, because the smart erase already checks if the blocks are empty.

File: src/sfdp.c [hunk 2]
```diff
@@ -143,6 +144,57 @@ static int read_jedec_flash_parameter(FLASHER_SPI_FLASH_T *ptFlash, unsigned lon
...
+		/** Read erase instruction entries from JEDEC Basic Flash Parametes
+		 *  Read starting from DWORD 8, there are 4 erase entries each with:
+		 *  8-Bit for the size in 2^N format
+		 *  8-Bit for the OpCode
+		 */
+		const unsigned char DWORD_8_START = 7*4; /* Nr of DWords starts at 1*/
+		unsigned int nrEraseOps = 0; /* Required for sorting */
+
+		/* Get erase operations from SFDP data */
+		for(int opNr = 0; opNr < FLASHER_SPI_NR_ERASE_INSTRUCTIONS; opNr++){
```
Comment (dizzydevil) [inline src/sfdp.c:156]: Do we have to take into account the case that there's a hole in the table, e.g. entries 0 and 2 are occupied?

Comment (jneis-at-hilscher) [reply src/sfdp.c:156]: You are correct in that there is no requirement for the fields to be filled in order, as far as I can tell.
There are already mechanisms to account for this:
- We always read all 4 instructions, if we find an empty instruction save it as 0 size, 0 opcode
- Sort the entries in ptFlash->tSpiErase ordered by erase size, starting with the smallest non-0 sized operation at element 0 (line 170)

However I just noticed that the sorting loop in line 171 will not sort correctly if the 0-Entries are in the middle.
I'll address this and get back to you.

File: templates/flasher.lua (hunk 1 above)
Comment (jneis-at-hilscher) [reply templates/flasher.lua:1013]: I didn't include this initially, however the new smart erase operation will fall back to the normal erase operation under various conditions:
1. We are not erasing the SPI-Flash but one of the other memories
2. There are less than 2 erase operations found in the erase operations array
3. The largest erase command is so large, we run out of memory to keep track of it (highly unlikely)

Point 2. is currently the case for most netX, as far as I can tell. This is because they get their flash information from the flasher-included table which currently only supports one sector erase and one page erase command.

I'd be open to removing this but I'm not certain if we should remove it, given the reason stated above.
What do you think?

Comment (dizzydevil) [reply templates/flasher.lua:1013]: I'm sorry, I didn't think of the fallback. It makes sense to keep the initial check. Maybe this can be optimized later on.

File: src/sfdp.c [hunk 3]
```diff
@@ -143,6 +144,57 @@ static int read_jedec_flash_parameter(FLASHER_SPI_FLASH_T *ptFlash, unsigned lon
...
+					tmpEraseOpCode = ptFlash->tSpiErase[eraseCmd].OpCode;
+					tmpMinPos = eraseCmd;
+				}
+			}
+			tmpEraseSize = ptFlash->tSpiErase[sortPos].Size;
+			tmpEraseOpCode = ptFlash->tSpiErase[sortPos].OpCode;
+			ptFlash->tSpiErase[sortPos] = ptFlash->tSpiErase[tmpMinPos];
+			ptFlash->tSpiErase[tmpMinPos].Size = tmpEraseSize;
+			ptFlash->tSpiErase[tmpMinPos].OpCode = tmpEraseOpCode;
+		}
```
Comment (dizzydevil) [inline src/sfdp.c:189]: This is more convoluted than it needs to be. It took me a while to understand.
How about 
FLASHER_SPI_ERASE_T tTmpErase;
unsigned int min_pos;
for (unsigned int pos1 = 0; pos1 < nrEraseOps; pos1++){
	min_pos = pos1;
	for (unsigned int pos2 = pos1+1; pos2 < nrEraseOps; pos2++){
		if (ptFlash->tSpiErase[pos2].Size < ptFlash->tSpiErase[min_pos].Size) {
			min_pos = pos2;
		}
	}
	
	tTmpErase = ptFlash->tSpiErase[pos1];
	ptFlash->tSpiErase[pos1] = ptFlash->tSpiErase[min_pos];
	ptFlash->tSpiErase[min_pos] = tTmpErase;
}

Or even plain bubblesort.

File: src/sfdp.c (hunk 2 above)
Comment (dizzydevil) [reply src/sfdp.c:156]: First clear the whole list, then copy only those entries that are non-zero:
	if(shift != 0 && shift < sizeof(unsigned long long)*8){
		ptFlash->tSpiErase[nrEraseOps].Size = 1U << shift;
		ptFlash->tSpiErase[nrEraseOps].OpCode = opCode;
		nrEraseOps++;
	}

File: src/flasher_spi.c [hunk 4]
```diff
@@ -537,6 +537,235 @@ NETX_CONSOLEAPP_RESULT_T spi_flash(const FLASHER_SPI_FLASH_T *ptFlashDescription
...
+ */
+NETX_CONSOLEAPP_RESULT_T spi_smart_erase(const FLASHER_SPI_FLASH_T *ptFlashDescription, const unsigned long ulStartAdr, const unsigned long ulEndAdr)
+{
+	/* Be Pessimistic */
+	NETX_CONSOLEAPP_RESULT_T retVal = NETX_CONSOLEAPP_RESULT_ERROR;
+	const SPIFLASH_ATTRIBUTES_T attributes = ptFlashDescription->tAttributes;
+	const FLASHER_SPI_ERASE_T *eraseTypes = ptFlashDescription->tSpiErase;
+	
+	/* Nr of valid entries */
+	unsigned int nrEraseOps = 0;
```
Comment (dizzydevil) [inline src/flasher_spi.c:623]: You could just store the number of valid erase operations when detecting them in the sdfp code.

File: src/flasher_spi.c [hunk 5]
```diff
@@ -537,6 +537,235 @@ NETX_CONSOLEAPP_RESULT_T spi_flash(const FLASHER_SPI_FLASH_T *ptFlashDescription
...
+		}
+	}
+
+	/* Fallback to normal erase due to lack of erase operations */
+	if(nrEraseOps < 2)
+	{
+		uprintf(". Less than two erase operations detected, using normal erase\n");
+		retVal = spi_erase(ptFlashDescription, ulStartAdr, ulEndAdr);
+		return retVal;
+	}
```
Comment (dizzydevil) [inline src/flasher_spi.c:638]: This check and the fallback could be made earlier, outside the smart erase function. The SFDP data are evaluated during the flash detect operation, so we already know if it makes sense to do a smart erase.

File: templates/flasher.lua (hunk 1 above)
Comment (dizzydevil) [reply templates/flasher.lua:1013]: About the fallback in case 2 (and 3?): Even if there's only one size of erase block available, don't we want to erase only those blocks that are not already empty?

File: lua/cli_flash.lua [hunk 6]
```diff
@@ -283,6 +283,22 @@ addJtagResetArg(tParserCommandErase)
...
+addUnitOptionArg(tParserCommandSmartErase)
+addChipSelectOptionArg(tParserCommandSmartErase)
+addStartOffsetArg(tParserCommandSmartErase)
+addLengthArg(tParserCommandSmartErase)
+-- optional_args = {"p", "t", "jf", "jr"}
+addPluginNameArg(tParserCommandSmartErase)
+addPluginTypeArg(tParserCommandSmartErase)
+addJtagResetArg(tParserCommandSmartErase)
+addJtagKhzArg(tParserCommandSmartErase)
+addSecureArgs(tParserCommandSmartErase)
```
Comment (dizzydevil) [inline lua/cli_flash.lua:299]: There's nothing wrong with this, but do we want an extra erase command, or rather an option to the normal erase command? What about the erase operation that may be done before flashing data?

File: src/flasher_spi.c [hunk 7]
```diff
@@ -537,6 +537,235 @@ NETX_CONSOLEAPP_RESULT_T spi_flash(const FLASHER_SPI_FLASH_T *ptFlashDescription
...
+		/* Find optimal erase pattern for read sectors */
+		/* Check each erase command, starting with the largest */
+		for(int eraseCmd = (int) nrEraseOps-1; eraseCmd >= 0; eraseCmd--)
+		{
+			unsigned int nrSectorsEraseable = eraseTypes[eraseCmd].Size/sectorSizeBytes;
+			/* Check if we found a page erase command, we only erase by sector */
+			if(nrSectorsEraseable < 1)
+			{
+				continue;
+			}
```
Comment (dizzydevil) [inline src/flasher_spi.c:714]: This means that the routine cannot erase with page granularity even on a flash device that supports page erase.
Why is the logic based on sector erase and not on the smallest erase block size?

The normal SPI erase routine, spi_erase_with_progress, can erase with page granularity.
However, it won't do so normally, because spi_getEraseArea always extends the area to be erased to sector bounaries.
So maybe it's no problem that we can't erase single pages.

File: lua/cli_flash.lua (hunk 6 above)
Comment (dizzydevil) [reply lua/cli_flash.lua:299]: Ok for now, may be turned into an option later on or even replace the standard erase.

**Review by dizzydevil — APPROVED** [review]

**Review by jneis-at-hilscher — COMMENTED** [review]
I ran some tests and found some edge cases where the smart_erase runs slower than the normal erase.
A change has been requested to always use the SFDP data unless specified otherwise by a flag. This will be implemented soon.

File: src/flasher_spi.c (hunk 7 above)
Comment (dizzydevil) [reply src/flasher_spi.c:718]: "My biggest concern is that we're only gaining a marginal improvement" - I wasn't thinking about a speed increase, but rather about being able to erase smaller areas, but as said, since get_erase_area expands the area to the next sector boundaries anyway and we didn't need page-granularity so far, I don't think it's important.

File: templates/flasher.lua (hunk 1 above)
Comment (dizzydevil) [reply templates/flasher.lua:1052]: On the one hand, it does not matter, since smart erase currently only works with SPI flash, on the other hand, we generally want to fall back to normal erase automatically if smart erase is not possible for some reason, so we'll keep the is_erased check.

File: src/main.c [hunk 8]
```diff
@@ -322,6 +325,56 @@ static NETX_CONSOLEAPP_RESULT_T opMode_erase(tFlasherInputParameter *ptAppParams
...
+			return tResult;
+		}
+		break;
+
+	case BUS_ParFlash:
+		/*  use parallel flash - not yet implemented.  */
+	case BUS_IFlash:
+		/*  use internal flash - not yet implemented.  */
+	case BUS_SDIO:
+		/*  use sdio flash - not yet implemented.  */
```
Comment (dizzydevil) [inline src/main.c:364]: If smart erase is not implemented, print a message "Smart erase is not implemented for parallel flash, using standard erase instead" and call normal erase.

File: templates/flasher.lua [hunk 9]
```diff
@@ -522,6 +524,11 @@ function M.detect(tPlugin, aAttr, tBus, ulUnit, ulChipSelect, fnCallbackMessage,
 		local ulMmioConfiguration = atParameter.ulMmioConfiguration
 		ulMmioConfiguration = ulMmioConfiguration or 0xffffffff
 
+		-- Set flag for use of SFDP erase operation if smart_erse command was used
+		if bUseSfdpErase then
+			ulFlags = ulFlags + 1
+		end
+
```
Comment (dizzydevil) [inline templates/flasher.lua:531]: Define a constant, like
FLAG_DETECT_SPI_USE_SFDP_ERASE = 1.
Also, with future extensions in mind, we might take a parameter ulFlags rather than having a separate parameter for every setting.

File: src/flasher_spi.c [hunk 10]
```diff
@@ -537,6 +537,251 @@ NETX_CONSOLEAPP_RESULT_T spi_flash(const FLASHER_SPI_FLASH_T *ptFlashDescription
...
+	for (unsigned int i = 0; i < SPI_ERASE_BITMAP_ENTRY_CNT; i++)
+	{
+		sectorBitmap[i] = 0;
+	}
+	
+	if(bitmapSize > SPI_ERASE_BITMAP_ENTRY_CNT){
+		uprintf("! Remembering too many memory sections required for largest erase command, falling back to normal erase\n");
+		retVal = spi_erase(ptFlashDescription, ulStartAdr, ulEndAdr);
+		return retVal;
+	}
```
Comment (dizzydevil) [inline src/flasher_spi.c:667]: If the largest erase block is too large to handle, we could still use the smaller block sizes.
I'd check the erase block sizes and remove any block sizes that are too small or too large, either  while reading the erase commands from SFDP or later in smart_erase.

File: src/main.c (hunk 8 above)
Comment (dizzydevil) [reply src/main.c:364]: Sorry, I didn't see the call to erase in the default case :)

**Comment by FrikadellenFred** [comment]
Pull request is closed. it was resolved by cherry picking it to a newer state of the project
